File: 프로그래머스/2/72412. 순위 검색/순위 검색.java
import java.util.*;

class Solution {
    public int[] solution(String[] info, String[] query) {
        Map<String, List<Integer>> infos = new HashMap<>();
        
        for (String inf : info) {
            String[] s = inf.split(" ");
            
            for (String o1 : List.of(s[0], "-")) {
                for (String o2 : List.of(s[1], "-")) {
                    for (String o3 : List.of(s[2], "-")) {
                        for (String o4 : List.of(s[3], "-")) {
                            String ls = o1 + o2 + o3 + o4;
                            Integer value = Integer.valueOf(s[4]);
                            infos.merge(ls, new ArrayList<Integer>(List.of(value)), (o, v) ->{
                                o.add(Integer.valueOf(value));
                                return o;
                                }
                            );
                        }
                    }
                }
            }
        }
        
        for (Map.Entry<String, List<Integer>> kv : infos.entrySet()) {
            kv.getValue().sort(Comparator.naturalOrder());
        }
        
        int[] answer = new int[query.length];
        for (int i = 0; i < query.length; i++) {
            int result = 0;
            String[] s = query[i].split(" ");
            String key = s[0] + s[2] + s[4] + s[6];
            if (infos.containsKey(key)) {
                result += findBigger(infos.get(key), Integer.valueOf(s[7]));
            }
            answer[i] = result;
        }
        
        return answer;
    }
    
    public int findBigger(List<Integer> scores, int score) {
     
        int left = 0;
        int right = scores.size() - 1;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            if (scores.get(mid) >= score) {
                right = mid - 1;
            } 
            else if (scores.get(mid) < score) {
                left = mid + 1;
            }
        }
        
        return scores.size() - left;
    }
}

File: 프로그래머스/2/72412. 순위 검색/순위 검색.py
from collections import defaultdict
from bisect import bisect_left
import re

def solution(infos, query):
    data = defaultdict(list)
    
    for info in infos:
        s_info = info.split(" ")
        
        for i in (s_info[0], "-"):
            for j in (s_info[1], "-"):
                for k in (s_info[2], "-"):
                    for l in (s_info[3], "-"):
                        data[(i,j,k,l)].append(int(s_info[4]))
    for key in data:
        data[key].sort()
    
    answer = []
    for q in query:
        s_q = re.split(" and | ", q)
        scores = data[tuple(s_q[:4])]
            
        answer.append(len(scores) - bisect_left(scores, int(s_q[4])))
    return answer